tests/e2e/test_operators.py

```diff
@@ -5,397 +5,454 @@
         capture_output=True,
         cwd=utils.BASE_DATA_PATH,
+        text=True,
+        env=env,
     )
 
-    output = utils.assert_playbook_output(result)
-
-    printed_events = [
-        line for line in output if "Event matched" in line["stdout"]
-    ]
-
-    assert len(printed_events) == expected_count
-
-
-@pytest.mark.e2e
-def test_not_operator():
+    assert result.returncode == 0
+    assert not result.stderr
+
+    with check:
+        assert (
+            "Event matched: {'id': 'test_contains_1'" in result.stdout
+        ), "contains operator failed"
+
+    with check:
+        assert (
+            len(
+                [
+                    line
+                    for line in result.stdout.splitlines()
+                    if "not-in #1 passed" in line
+                ]
+            )
+            == 2
+        ), "not-in operator failed hardcoded list"
+
+    with check:
+        assert (
+            "not-in #2 passed" in result.stdout
+        ), "not-in operator failed, string"
+
+    with check:
+        assert (
+            len(
+                [
+                    line
+                    for line in result.stdout.splitlines()
+                    if "not-in #3 passed" in line
+                ]
+            )
+            == 1
+        ), "not-in operator failed, integer"
+
+    with check:
+        assert (
+            len(
+                [
+                    line
+                    for line in result.stdout.splitlines()
+                    if "not-in #4 passed" in line
+                ]
+            )
+            == 1
+        ), "not-in operator failed, mixed types"
+
+    with check:
+        assert (
+            "Winter is missing" not in result.stdout
+        ), "not-cointains operator failed, output found"
+
+    with check:
+        assert (
+            "not-contains #2 passed" in result.stdout
+        ), "not-contains operator failed"
+
+    with check:
+        assert (
+            "not-contains #3 passed" in result.stdout
+        ), "not-contains operator failed"
+
+    with check:
+        assert (
+            len(
+                [
+                    line
+                    for line in result.stdout.splitlines()
+                    if "test in operator #1 passed" in line
+                ]
+            )
+            == 1
+        ), "in operator failed, string hardcoded list"
+
+    with check:
+        assert (
+            len(
+                [
+                    line
+                    for line in result.stdout.splitlines()
+                    if "test in operator #2 passed" in line
+                ]
+            )
+            == 4
+        ), "in operator failed"
+
+    with check:
+        assert (
+            len(
+                [
+                    line
+                    for line in result.stdout.splitlines()
+                    if "test contains operator #1 passed" in line
+                ]
+            )
+            == 2
```

Comment:
There are four rules for the "contains" operator, but only two are matching. The third is not matching because it's looking for 6.99, but the list contains 6.90. The fourth is trying to match an empty list with an empty string but this doesn't seem to work.